sdk/core/System.ClientModel/src/Internal/ClientModelEventSource.cs

```diff
@@ -0,0 +1,318 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics.Tracing;
+using System.Text;
+
+namespace System.ClientModel.Internal;
+
+[EventSource]
+internal sealed class ClientModelEventSource : EventSource
+{
+    private const int RequestEvent = 1;
+    private const int RequestContentEvent = 2;
+    private const int ResponseEvent = 5;
+    private const int ResponseContentEvent = 6;
+    private const int ResponseDelayEvent = 7;
+    private const int ErrorResponseEvent = 8;
+    private const int ErrorResponseContentEvent = 9;
+    private const int RequestRetryingEvent = 10;
+    private const int ResponseContentBlockEvent = 11;
+    private const int ErrorResponseContentBlockEvent = 12;
+    private const int ResponseContentTextEvent = 13;
+    private const int ErrorResponseContentTextEvent = 14;
+    private const int ResponseContentTextBlockEvent = 15;
+    private const int ErrorResponseContentTextBlockEvent = 16;
+    private const int RequestContentTextEvent = 17;
+    private const int ExceptionResponseEvent = 18;
+    private const int BackgroundRefreshFailedEvent = 19;
+    private const int RequestRedirectEvent = 20;
+    private const int RequestRedirectBlockedEvent = 21;
+    private const int RequestRedirectCountExceededEvent = 22;
+    private const int PipelineTransportOptionsNotAppliedEvent = 23;
+
+    private ClientModelEventSource(string eventSourceName, string[]? traits) : base(eventSourceName, EventSourceSettings.Default, traits) { }
+
+    public static ClientModelEventSource Create(string eventSourceName, string[]? traits) => new(eventSourceName, traits);
```

Comment:
IIRC, there was offline discussion with the System.Diagnostics team about edge cases where this approach would fail, and it being OK to document that these cases are unsupported.  If it's not already here somewhere, it would be good to make sure the supported and unsupported cases are written down in code comments for future maintainers, perhaps more verbosely than what we put in refdocs for end-users.

Reply:
Yes I will get that added